Implements GEP 709 - ReferencePolicy (#741)
kubernetes-sigs/gateway-api: site-src/v1alpha2/references/cross-namespace-references.md

# Cross namespace references and ReferencePolicy

> ### Terminology note

Reviewer: This feels a bit out of place. Not sure how this would render with mkdocs, but maybe something like a blockquote or callout format with this content added in the Intro section right before/after "Referent" is mentioned for the first time.

Author: Yeah, I didn't like it there either, but couldn't think of the right place to put it. I'll see how we can do a blockquote or something more inline instead.

Author: I think I found a "Note" syntax that should work, I'll check the deploy preview.

Author: I added in an "admonitions" extension, that we can use with:

```
!!! admonition
    blockquote text is indented by four spaces
```

Reviewer: This looks perfect! Thanks for adding it in

When discussing the process of creating cross-namespace object references, this
document and the documentation on the API itself refers to the object being
referred to as "the referent object", using the meaning of "referent" to be

> "the person, thing, or idea that a word, phrase, or object refers to".[1](https://dictionary.cambridge.org/dictionary/english/referent)

Reviewer: This footnote looks a bit off in the deploy preview. Maybe either use `<sup>` tag or a full word like `(source)`.
https://deploy-preview-741--kubernetes-sigs-gateway-api.netlify.app/v1alpha2/references/cross-namespace-references/

for the details of the object's behavior.

### Exceptions
There are some situations where it MAY be acceptable to ignore ReferencePolicy
in favor of some other security mechanism. This MAY only be done if other
mechanisms like NetworkPolicy can effectively limit cross-namespace references
by the implementation.

An implementation choosing to make this exception MUST clearly document that
ReferencePolicy is not honored by their implementations and detail which
alternative safeguards are available. Note that this is unlikely to apply to
ingress implementations of the API and will not apply to all mesh
implementations.

For an example of the risks involved in cross-namespace references, refer to
[CVE-2021-25740](https://github.com/kubernetes/kubernetes/issues/103675).
Implementations of this API need to be very careful to avoid confused deputy
attacks. ReferencePolicy provides a safeguard for that. Exceptions MUST only
be made by implementations that are absolutely certain that other equally
effective safeguards are in place.

### Conformance Level

ReferencePolicy support is a "CORE" conformance level requirement for

> cross-namespace references inside the following objects:

Reviewer 2: ```suggestion
cross-namespace references that originate from the following objects:
```

- HTTPRoute
- TLSRoute
- TCPRoute
- UDPRoute

That is, all implemenations MUST use this flow for any cross namespaces
in any of the core xRoute types, except as noted in the Exceptions section
above.

Other "ImplementationSpecific" objects and references MUST also use this flow
for cross-namespace references, except as noted in the Exceptions section

> above.

Reviewer: More a note for me, but when I implement the Route-Gateway binding GEP, I'll need to update this to discuss the uniqueness of that relationship.

# Cross namespace references and ReferencePolicy

In the Gateway API, it is possible to have references between objects cross namespace boundaries.
In particular, it's expected that Gateways and Routes may exist in different namespaces,

> or that Services may be referred to by Routes in a another namespace.

Reviewer 2: > that Gateways and Routes may

ReferencePolicy is not expected to be applied to Gateway-Route references so it would be best to avoid mentioning that here.

However, this significantly violates the idea of a namespace as the edge of a trust domain.
In order to bring cross-namespace references under the control of the owner of the referent object's namespace,

> the Gateway API has a ReferencePolicy object that must be created in the referent namespace for the reference to be successful.

Reviewer 2: nit: 80 char limit throughout the file

Reviewer 2: > object that must be created in the referent namespace

I know we use 'referent' a lot and assume that it is understood by our users but I think explaining the referent and referee would help here. This paragraph is very hard to parse for me even though I'm familiar with the concept.

Author: Added a note on terminology at the start, I think that's a great idea.

To put this another way, if an object is referred to from outside its namespace,
the object's owner must create a ReferencePolicy object that describes how that reference is allowed.
This page explains how this process works.

In the past, we've seen that forwarding traffic across namespace boundaries is a desired feature,
but without the kinds of safeguards proposed here, [vulnerabilities](https://github.com/kubernetes/kubernetes/issues/103675)

> can emerge.

Reviewer 2: Maybe start with this paragraph to build up to the problem we are trying to solve and then propose the solution.

## ReferencePolicy

To ensure that Gateway API is able to safely provide this functionality,
we need to enforce a handshake mechanism that requires resources in both namespaces to agree to this reference.
To accomplish that, a ReferencePolicy resource has been be introduced.

![Reference Policy](images/referencepolicy.png)

> With this model, Routes are able to directly reference Routes and Services in other namespaces.

Reviewer 2: We don't have Route->Route yet so it would be best not to mention it here. Can spring confusion.

These references are only considered valid if a ReferencePolicy in the target namespace explicitly allows it.

The following example shows how a HTTPRoute in namespace `foo` can reference a Service in namespace `bar`.
In this example a ReferencePolicy in the `bar` namespace explicitly allows references to Services from HTTPRoutes in the `foo` namespace.

```yaml
kind: HTTPRoute
metadata:
  name: foo
  namespace: foo
spec:
  rules:
  - matches:
    - path: /bar
    forwardTo:
      backend:
      - name: bar
        namespace: bar
---
kind: ReferencePolicy
metadata:
  name: bar
  namespace: bar
spec:
  from:
  - group: networking.gateway.k8s.io
    kind: HTTPRoute
    namespace: foo
  to:
  - group: core
    kind: Service
```

### API design decisions

> This proposed API is fairly straightforward, but comes with a few notable decisions:

Reviewer 2: nit:
```suggestion
While the API is simplistic in nature, it comes with a few notable decisions:
```

Avoid blanket statements that the user might not grasp.

Author: Thanks, that is a bad habit of mine.

1. Each ReferencePolicy only supports a single From and To section.
   Additional trust relationships must be modeled with additional ReferencePolicy resources.
1. Resource names are intentionally excluded from this policy for simplicity and because they rarely provide any meaningful protection.
   A user that is able to write to resources of a certain kind within a namespace can always rename
   resources or change the structure of the resources to match a given policy.

> 1. A single Namespace is allowed per "From" struct.

Reviewer 3: is "all"/"any" not a common use case? I see the concerns of making it too permissive but generally APIs offer some way to be fairly permissive, even if its not easy.

Reviewer 2: +1 to both the comments above

>    Although a selector would be more powerful it may encourage unnecessarily insecure configuration.

Reviewer 2: I don't mind being more assertive here and seeing how it plays out in the community.
```suggestion
   Although a selector would be more powerful, it encourages unnecessarily insecure configuration.
```

> ```go

Author: Yeah, good idea. Done.

    //
    // * HTTPRoute
    // * TCPRoute
    // * TLSRoute
    // * UDPRoute
    //
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=253
    Kind string `json:"kind"`

    // Namespace is the namespace of the referent.
    //
    // Support: Core
    //
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=253
    Namespace string `json:"namespace,omitempty"`
}

// ReferencePolicyTo describes what Kinds are allowed as targets of the
// references.
type ReferencePolicyTo struct {
    // Group is the group of the referent.
    //
    // Support: Core
    //
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=253
    Group string `json:"group"`

    // Kind is the kind of the referent. Although implementations may support
    // additional resources, the following types are part of the "Core"
    // support level for this field:
    //
    // * Service
    // * HTTPRoute
    // * TCPRoute
    // * TLSRoute
    // * UDPRoute
    //
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=253
    Kind string `json:"kind"`
}
```

> ### Benefits

Reviewer: I think this section is also unnecessary for the docs.

	//
	// Support: Core
	//
	// +kubebuilder:validation:MinLength=1
	// +kubebuilder:validation:MaxLength=253
	// +optional
	Namespace *string `json:"namespace,omitempty"`
}

// From httproute_types.go

	// BackendRef is a reference to a backend to forward matched requests to. If
	// both BackendRef and ServiceName are specified, ServiceName will be given
	// precedence.
	//
	// If the referent cannot be found, the route must be dropped
	// from the Gateway. The controller should raise the "ResolvedRefs"
	// condition on the Gateway with the "DegradedRoutes" reason.
	// The gateway status for this route should be updated with a
	// condition that describes the error more specifically.
	//
	// If there is a cross-namespace reference to an *existing* object
	// with no ReferencePolicy, the controller must ensure the "ResolvedRefs"
	// condition on the Gateway is set to `status: true`, with the "RefNotPermitted"
	// reason.
	//
	// Support: Custom
	//
	// +optional
	BackendRef *ObjectReference `json:"backendRef,omitempty"`

    // The BackendRef in the more general RouteForwardTo object in shared_types.go
    // has also been updated.
```
### Conformance Level

> ReferencePolicy support is a "CORE" coformance level for the following

Reviewer 4: ```suggestion
ReferencePolicy support is a "Core" conformance level for the following
```

Reviewer: I'm not sure which is better here, but we should be consistent. Our [docs around conformance levels](https://gateway-api.sigs.k8s.io/concepts/guidelines/#conformance) use all caps, but usually the spec does not. Do we have a strong preference one way or the other?

Author: I lifted this phrasing from one of the other GEPs - I don't mind changing, but if so, we should change the others as well.

objects:
- HTTPRoute
- TLSRoute
- TCPRoute

> - UDPRoute

Reviewer 2: I find this conformance guidance is overly broad and vague.
We should be more explicit here: XRoute -> cross-namespace service/secrets require RP resource.

Author: I've clarified a bit, does this read better?

> Other "ImplementationSpecific" objects and references are *strongly recommended* to also use this flow for cross-namespace references. 

Reviewer 2: We should use "must" here rather than a recommendation.
I do understand that we can't enforce it.
A "recommendation" could be neglected as opposed to "must".
That also helps with security issues that come down in future.

Author: Agreed, I made it a recommendation precisely because we can't really enforce it, but putting it as a MUST is better.
